internal/acctest/spauthenticationpolicycontractmapping/sp_authentication_policy_contract_mapping_test.go

```diff
@@ -0,0 +1,317 @@
+package acctest_test
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-framework/providerserver"
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	"github.com/hashicorp/terraform-plugin-testing/terraform"
+	client "github.com/pingidentity/pingfederate-go-client/v1125/configurationapi"
+	"github.com/pingidentity/terraform-provider-pingfederate/internal/acctest"
+	"github.com/pingidentity/terraform-provider-pingfederate/internal/provider"
+)
+
+const spAuthenticationPolicyContractMappingId = "authenticationpolicycontractid|spadapter"
+const apcSourceId = "authenticationpolicycontractid"
+const spTargetId = "spadapter"
+
+type spAuthenticationPolicyContractMappingResourceModel struct {
+	attributeSource              *client.JdbcAttributeSource
+	attributeContractFulfillment client.AttributeFulfillmentValue
+	issuanceCriteria             *client.ConditionalIssuanceCriteriaEntry
+	sourceId                     string
+	targetId                     string
+}
+
+func stringPointer(val string) *string {
+	return &val
+}
+
+func initialAttributeContractFulfillment() client.AttributeFulfillmentValue {
+	initialAttributecontractfulfillment := *client.NewAttributeFulfillmentValue(
+		*client.NewSourceTypeIdKey("TEXT"),
+		"value",
+	)
+	return initialAttributecontractfulfillment
+}
+
+func updatedAttributeContractFulfillment() client.AttributeFulfillmentValue {
+	updatedAttributecontractfulfillment := *client.NewAttributeFulfillmentValue(
+		*client.NewSourceTypeIdKey("CONTEXT"),
+		"ClientIp",
+	)
+	return updatedAttributecontractfulfillment
+}
+
+func attributeContractFulfillmentHclBlock(aCf *client.AttributeFulfillmentValue) string {
+	var builder strings.Builder
+	if aCf == nil {
+		return ""
+	}
+	if aCf != nil {
+		builder.WriteString("      source = {\n")
+		builder.WriteString("        type = \"")
+		builder.WriteString(aCf.Source.Type)
+		builder.WriteString("\"\n")
+		builder.WriteString("      },\n")
+		builder.WriteString("      value = \"")
+		builder.WriteString(aCf.Value)
+		builder.WriteString("\"")
+	}
+	return builder.String()
+}
+
+func attributeSource() *client.JdbcAttributeSource {
```

Comment:
Same for attribute sources- this should be available in a common test package, no need to redefine it